### PR TITLE
Add functions for layout properties

### LINKED ISSUE #361: Functions for layout properties
Currently bucket rules can only be static, which means attributes important for point and label placement (`fontSize`, `padding`, `size`, `spacing`, etc.) are set to a single value through the whole zoom range.

Considering that buckets are parsed individually for each tile, which has a zoom value, we can introduce the same functions we use for other rules to bucket rules, with the only difference that the values will be calculated based on tile zoom as opposed to arbitrary fractional zoom. It is also important in case we ditch buckets in user styles, to be consistent.

This issue extends #317

### PR DESCRIPTION
**Pull request by lbud**
Closes #361.

- [x] Figure out array properties (`text-offset` and `icon-offset`) — right now they're hitting some bug in style_declaration.js
- [x] Add tests

### CODE REVIEW

File: js/data/create_bucket.js [hunk 1]
```diff
@@ -15,6 +16,10 @@ function createBucket(layer, buffers, collision, indices) {
         return;
     }
 
+    for (var k in layer.layout) {
+        layer.layout[k] = new StyleDeclaration('layout', layer.type, k, layer.layout[k]).value;
```
Comment (jfirebaugh) [inline js/data/create_bucket.js]: `layer.layout[k] = new StyleDeclaration('layout', layer.type, k, layer.layout[k]).calculate(z)`

Pass the tile zoom as `z` into `createBucket`. Then you should be able to revert all the changes to the various individual bucket classes.

Comment (lbud) [inline js/data/create_bucket.js]: I'm trying this, but when it's calculated in create_bucket for some reason they don't get reevaluated at different zoom levels — i.e. with a style with
``` json
"text-rotate": {
        "stops": [[14,0], [16, 190]]
      }
```
if you load the page at zoom 18, the text is rotated, but if you load it at 14 it isn't — and never flips. Newly added labels *are* rotated at the z-level where they first come in. … :confused: ?

Comment (jfirebaugh) [inline js/data/create_bucket.js]: Hmm. When you load it at z14 and then zoom in, does it load z15+ tiles?

Comment (lbud) [inline js/data/create_bucket.js]: Ah I see. Nope, worker_tile only ever knows about max z15 — so should I be threading the map's zoom through into the worker?

Comment (jfirebaugh) [inline js/data/create_bucket.js]: Oh, right, z14 is the max for most VT sources. So the bug is that it appears rotated at z18. It's only ever loading z14 tiles, so only the first stop value should be used.

Comment (lbud) [inline js/data/create_bucket.js]: Hm, no, even when I change those z-levels to 10 and 13, they never recalculate on zoom using this method…

Comment (lbud) [inline js/data/create_bucket.js]: nvm figured this out

**Comment by lbud** [comment]
The big weird issue here is that functions only work up to z15 (in streets, for example) because of actual vector tile zoom levels. Is that just a caveat we put in the spec?

**Comment by mourner** [comment]
@lbud yeah, I think it's all right for now. We can revisit this later.

**Comment by jfirebaugh** [comment]
Yeah, if necessary we can recalculate z15, 16, etc geometries with z14 tile data, but let's wait to see if that's needed.

**Pull request merged**